calico_containers/adapter/datastore.py

```diff
@@ -716,6 +792,51 @@ def get_ep_id_from_cont(self, hostname, container_id):
                 "Container with ID %s has no endpoints." % container_id)
 
     @handle_errors
+    def query_endpoint(self, endpoint_id=None, workload_id=None, orchestrator_id=None, hostname=None):
+        """
+        Optimized function to get endpoint(s).
+
+        Constructs a etcd-path that it as specific as possible given the provided criteria, in order
+        to return the smallest etcd tree as possible. After querying with the ep_path, it will then
+        compare the returned endpoints to the provided criteria, and return all matches.
+
+        :param endpoint_id: The ID of the endpoint
+        :param hostname: The hostname that the endpoint lives on.
+        :param workload_id: The workload (or container) that the endpoint belongs to.
+        :param orchestrator_id: The workload (or container) that the endpoint belongs to.
+
+        :return: A list of Endpoint Objects which match the criteria, or an empty list if none match
+        """
+        # First build the query string as specific as possible. Note, we want the query to be as specific as
+        # possible, so we proceed any variables with known constants e.g. we add '/workload' after the hostname
+        # variable.
+        ep_path = "/calico/v1/host"
+        if hostname:
+            ep_path += "/%s/workload" % hostname
+            if orchestrator_id:
+                ep_path += "/%s" % orchestrator_id
+                if workload_id:
+                    ep_path += "/%s/endpoint" % workload_id
+                    if endpoint_id:
+                        ep_path += "/%s" % endpoint_id
+        try:
+            # Search etcd
+            leaves = self.etcd_client.read(ep_path, recursive=True).leaves
+
+            # Filter through result
```

Comment:
move this outside the try.

in general it's good practise to minimize the code inside the try block to just that code that will actually raise the exceptions you are catching.